Support python 2 print statements in SConscripts

merlin66 opened the pull request:
This fixes a regression introduced in scons-3.0.0, where
SConscripts containing python 2 print statements would cause
syntax errors even when executing scons with python 2.7.

This ensures backward compatibility, allowing users to build
legacy code with scons-3.0.0 without having to patch it.

merlin66 commented:
Very happy with scons-3.0.0 so far, this is the only minor issue I found when building at my day job. This is my first pull request for scons, sorry if I missed something or should have done additional things. If you want me to add a test for this, I'm going to need some pointers.

timj commented on src/engine/SCons/Script/SConscript.py:
```diff
@@ -465,15 +463,15 @@ def EnsureSConsVersion(self, major, minor, revision=0):
         """Exit abnormally if the Python version is not late enough."""
         if sys.version_info < (major, minor):
             v = sys.version.split()[0]
-            print("Python %d.%d or greater required, but you have Python %s" %(major,minor,v))
```
For me this is a valid Python2 `print` statement so I'm not sure what is causing the problem.
```
Python 2.7.13 (default, Dec 18 2016, 07:03:39) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> print("hello")
hello
>>> print("hello %s" % (1,))
hello 1
>>> 
```
 Is it the `__future__` that's the real issue here? What error do you get? Is the problem that *really* old Python versions don't work with "print tuple"?

merlin66 replied:
The real issue is that I have an SConstruct that says this:
>> print "debug statement"

And even though this is valid code under python2, I can no longer build this with scons-3.0.0. My fix makes it build again, so it's important for backward compatibility.

timj replied:
My main worry now is that 3.0.0 is out and if we stop the `__future__` import for 3.0.1 then anyone who has "fixed" their `SConstruct` files and is using 2.7 will now see breakage.

merlin66 replied:
I just tested with python-2.7.13, and there you can write print('hello') without importing print_function. So removing the __future__ import should not break builds that change from print "hello" to print("hello") due to scons-3.0.0, at least in python-2.7.

This means the print statements also don't need to be changed, but only for this python version. For older python versions it might be needed, and for those python versions we will also have the breakage you are concerned about, if this pull request goes in.

merlin66 replied:
python-2.6.6 also supports both syntaxes without importing from __future__.

timj replied:
I don't think SCons works on Python 2.6 so I don't think that's an issue. The "print tuple" for mis okay so if people migrate by doing that then it's fine. The issue comes if they are printing to file handles where the syntax changed. @dirkbaechle  or @russel will have more ideas on how to proceed though.

timj commented:
If the problem is that the `__future__` import is changing the behavior of all the user `SConscript` files then I think that was an oversight. I recall we had a long debate on this and decided we should not contaminate user files with `print_function` if they were only ever using python2.

bdbaddog commented:
Can you create a simple test case (which gets skipped with python 3) to go along with this? 
(The idea is to verify that print "blah" works with py2 and that it is not broken going forward unintentionally)

Also please add a blurb in src/CHANGES.txt

merlin66 commented:
I will add a test, may need some hints to knock it into shape.

merlin66 commented:
I believe I have updated the pull request now. Reduced the patch to just removing the `__future__` import, and added a test: the first part of it is skipped with python 3, the other part is run on all python versions. I couldn't figure out how to run the tests, so I have not done that yet.

bdbaddog commented:
python runtest.py test/print_statement.py  should do the trick.
(python3.5 ... to try with py3.5.. etc)

merlin66 commented:
Fixed some errors in the test, and it now runs and passes both with python 2.7 and python 3.6. I also confirmed that it fails without my fix. It tests that both the python 2 style and python 3 style print statements work with python 2.7. When run with python 3 it only tests the `print("")` syntax.

How does this look?

bdbaddog commented:
I made a minor fix to your test. test.skip_test() wasn't warranted but rather test.pass if running under python3. As if it passed that far as far as py3 is concerned it passed.  I've merged and then cherry picked the changes into the new repo I'm prepping to fix issue with missing history caused by issues with github's mercurial conversion logic.
